cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/ETLSparkProgram.java

```diff
@@ -140,32 +131,104 @@ public Object call(Tuple2<String, Object> input) throws Exception {
     }
   }
 
+  private JavaPairRDD<String, Object> doTransform(SparkContext context, BatchPhaseSpec phaseSpec,
+                                                  JavaPairRDD<Object, Object> input,
+                                                  String aggregatorName, int numPartitions) throws Exception {
+
+    Set<StageInfo> sparkTransforms = phaseSpec.getPhase().getStagesOfType(SparkTransform.PLUGIN_TYPE);
+    if (sparkTransforms.isEmpty()) {
+      // if this is not a phase with spark transform, do regular transform logic
+      if (aggregatorName != null) {
+        JavaPairRDD<Object, Object> preGroupRDD =
+          input.flatMapToPair(new PreGroupFunction(context, null, aggregatorName));
+        JavaPairRDD<Object, Iterable<Object>> groupedRDD =
+          numPartitions < 0 ? preGroupRDD.groupByKey() : preGroupRDD.groupByKey(numPartitions);
+        return groupedRDD.flatMapToPair(new MapFunction<Iterable<Object>>(context, null, aggregatorName)).cache();
+      } else {
+        return input.flatMapToPair(new MapFunction<>(context, null, null)).cache();
+      }
+    }
+
+    // otherwise, special casing for SparkTransform type:
+
+    // there should only be no other plugins of type Transform, because of how Smart Workflow breaks up the phases
+    Set<StageInfo> stagesOfType = phaseSpec.getPhase().getStagesOfType(Transform.PLUGIN_TYPE);
+    Preconditions.checkArgument(stagesOfType.isEmpty(),
+                                "Found non-empty set of transform plugins when expecting none: %s", stagesOfType);
+
+    // Smart Workflow should guarantee that only 1 spark transform exists per phase. This can be improved in the future
+    // for efficiency.
+    Preconditions.checkArgument(sparkTransforms.size() == 1, "Expected only 1 spark transform: %s", sparkTransforms);
```

Comment:
nit: shouldn't happen, but lets put the transforms that ended up in the this stage in the message